### ISSUE TITLE
merge, join, fusion… functions for Tables

### ISSUE THREAD

**Issue by moebio** [moebiolabs/moebio_framework#240]
I will need some help from @protozoo to have some clarity on which operations, with what options, and under which names these currently exist in coding languages, libraries, and specially sql and SAP.

Note: some are already, although very likely poorly parametrized and named, part of mo, see:

TableOperators.mergeDataTables
TableOperators.mergeDataTablesList
TableOperators.sumDataTables

**Comment by Jeff-Clark** [comment]
newly added TableOperators.concatRows might be relevant also.

**Comment by protozoo** [comment]
Oh @Jeff-Clark, TableOperators.concatRows was requested SO often during the workshop... thanks.

@moebio, my understanding is that different libraries/languages tackle the subject in different manner.

## SQL
SQL uses the "join" concept. There are different ways of joining tables:

- INNER JOIN: Returns all rows when there is at least one match in BOTH tables
- LEFT JOIN: Return all rows from the left table, and the matched rows from the right table
- RIGHT JOIN: Return all rows from the right table, and the matched rows from the left table
- FULL JOIN: Return all rows when there is a match in ONE of the tables

However, most sources seem to agree that INNER JOIN and LEFT JOIN are the most used and cover almost all cases.

References:
- Quick explanation: http://www.w3schools.com/sql/sql_join.asp
- Visual explanation using Venn diagrams: http://blog.codinghorror.com/a-visual-explanation-of-sql-joins/

## SAP
I have no experience with SAP but looks like they also use the "join" concept (and "union", which seems to be analogous to SQL's "FULL JOIN")

Reference:
- SAP docs.: https://help.sap.com/saphelp_nw73/helpdata/en/5e/13a3d8fd42487ea5adee7d9379c34f/content.htm

## R
The R language uses the "merge" concept, which interestingly matches the way most people at the workshop referred to this action of bringing different tables together. In R, "merge" is a function accepting different options to configure the details of the action. A quick look seems to indicate that these options, combined, allow the same set of joins or merges from SQL and SAP. I find the option names (the argument names, actually) a bit cryptic, as I usually find in R. This may be just a personal glitch, however.

Reference:
- R's "merge" docs: https://stat.ethz.ch/R-manual/R-devel/library/base/html/merge.html

## Pandas
Pandas seem to aggregate all the above together, offering both merge, join and concatenate actions.

- Panda's docs: http://pandas.pydata.org/pandas-docs/stable/merging.html

___

I'm not sure which way we should go. On one side I think we should try to simplify as much as possible, but some operations are inherently complex and I don't think we want to drop functionality for the sake of simplicity...

What do you think?

**Comment by moebio** [comment]
@Jeff-Clark  @protozoo  I think:

- this benchmark research is great

- javascript and also mo/lichen already contain certain more or less substantial logics (also in naming) that perhaps will guide us at least in certain decisions 

- we should meet the three of us and have a brief brainstorm on this

**Comment by moebio** [comment]
@Jeff-Clark @protozoo 

It seems that in all these cases the operations are done when identifying a single variable that is present in both tables (and that in general share some values).

For these cases I think it's gonna be easy to develop a solution for two tables, and also for a list of tables, and that at least cover inner and left.


Dani, do you remember if in the workshop someone has a need of doing something more complex, involving more than just one variable that acts as a hinge? Jeff, have you ever seen a situation like that one?

**Comment by protozoo** [comment]
Yes, indeed. One may want to join multiple tables. Some people asked for it during the workshop and it's a common request in SQL. 

Please see this example using 3 tables (one of them describing relationships with the other two):
`#/dani/examples/sql_multiple`

![image](https://cloud.githubusercontent.com/assets/1389973/15053858/e0c9aea2-1304-11e6-8032-93841a4ecdc1.png)

**Comment by moebio** [comment]
@Jeff-Clark @protozoo 

Also, there are options regarding the selection of lists. On SQL you can either select all (*) in which case it will basically all add columns from one table and ll from the other, repeating the common lists.

What I think I'll do is not allowing for any selection, adding thus all columns, but not repeating the common list, and instead placing it on the beginning

B, A, C               Q, P, A, R

the join by A and A will produce:

A, B, C, Q, P, R

The reason is that, contrary to what happens in SQL, the user can easily filter/assemble the tables she wants before the join (equivalent to perform the selection before).

If you see any caveat here, please comment.

**Comment by moebio** [comment]
@protozoo Yes, joining multiple tables is common and will be done. But what I'm ask is about joining by more than one common variable (as when in the workshop some asked about aggregating by more than one variable).

**Comment by protozoo** [comment]
I'm not sure I get that, @moebio . Could you provide an example?

**Comment by moebio** [comment]
@protozoo I'll take your question as an answer (no, there doesn't seem to be a necessity -or even it hasn't ever be defined- of joining two or more tables by more than one key feature)

It's actually difficult to come up with an example (and I never had to do such kind of joining)


id  name   country       computer                 id   name     country     car
1   santi    colombia      mac                        3     santi      chile         bmw
2   dani     spain            mac                        5     dani       spain        vw
3    jeff      colombia      pc                           9      jeff      canada       fiat
4    jeff      canada        mac

imagine joining these two tables by t0.name = t1.name and t0.country = t1.country

One option is to ask for respect pairs, in which case an inner join will produce:

id  name   country       computer   car
4    jeff      canada        mac          fiat

Other is not do so and will produce a longer table, it will find each coincidence of name and complete, and then each coincidence of country and complete

id  name   country       computer      car 
1   santi    colombia      mac          bmw
3     santi      chile         mac         bmw                  
…

(way too complicate)

**Comment by protozoo** [comment]
@moebio @Jeff-Clark 

Not sure if this is exactly what you were asking, but I think it can be helpful to think about it in these terms:

The merge and aggregate actions are different steps, being executed sequentially.

1. Merge: Takes 2 or more tables and, based on shared fields and combination rules, generates a new one.
2. Aggregate: Involves only 1 table and reduces it based on aggregation or grouping criteria.

For step 1 I don't currently see a need for more than 1 shared field between any two tables.

For step 2 I do indeed see the need of multiple variables. An example would be a table with these columns:

- year
- month
- date
- week_day
- total_sales

As an analyst I may want to aggregate based on **year** and **week_day** so that I get 1 row for each combination of **year / week_day**. 

Does it help somehow?

**Comment by moebio** [comment]
@protozoo aggregateTable already aggregates by more than one variable. 

I don't think single join + aggregation produces the same result as joining by multiple key variables (certainly not if the join is not inner), but again, joining by more than one key variable seems too esoteric so we won't do it (until the unlikely event someone badly needs it).

**Comment by moebio** [comment]
@Jeff-Clark @protozoo added joinTwoTables, only INNER and LEFT modes (for the time being)

you can try it: santiago/tests/SQLonTable

**Comment by Jeff-Clark** [comment]
joinTwoTables looks useful for sure. Currently the help in Lichen looks like this:
![image](https://cloud.githubusercontent.com/assets/1721620/15073725/3283de3e-1369-11e6-8698-0dbd47bbd96b.png)

But the code parm block in TableOperators.js like this:
![image](https://cloud.githubusercontent.com/assets/1721620/15073767/66998318-1369-11e6-8b09-6522e7674d07.png)

It looks like the framework_info.json checked in doesn't match the source and was generated from a version that has another mode 'full outer'.

For the parm descriptions it might be good to stick to our nomenclature, use 'list' or 'column' rather than 'variable'. Also it would be beneficial to describe what the modes do rather than just using the SQL terminology. People who don't know SQL likely don't know what is mean't by 'inner' or 'left join' etc.

**Comment by moebio** [comment]
@Jeff-Clark thanks for the feedback. Yes the inlet tooltip is wrong in two independent ways, I'm gonna fix this (btw I tried an outer and a full versions, but found it quite unnecessary… their phantoms remained…)

The reason I write:

index (Number) or name (String) of variable that is also present in table1

and not:

index (Number) or name (String) of list that is also present in table1

is because the variable might be the same in both tables, whereas one can't say the same for lists, they are different.

What do you think of this alternative:

index (Number) or name (String) of list that is also represented by another list (same variable) in table1

?

Also agree with expanding explanation on modes. I'll try something and then feel free to edit it. Explain this combinatoric process in a short text is quite challenging… but I agree that at least part of the process should be explained.

**Comment by Jeff-Clark** [comment]
@moebio I think your alternative is better. Agree about difficulty of explaining the modes succinctly.

**Comment by moebio** [comment]
@protozoo I need help here, I do not have an understanding of what`s going on here in your example:

SELECT 
t0.name as personName, 
t1.name as country
FROM t2
LEFT JOIN t0
    on t2.id_person = t0.id
LEFT JOIN t1
    on t2.id_country = t1.id

Is this a succession of two joins, one after the other?
-
More in general, when one joins more than two tables, the expected result is the accumulation of joins:
(((A join B) join C) join D)
?

**Comment by protozoo** [comment]
I guess it is, yes. 
It will join tables t2 and t0 based on the first criteria (t2.id_person = t0.id) and apply the same operation with t1. 

Not sure if the join order is relevant in terms of results (apparently it dos not), but looks like there is in terms of performance/optimization, see [this link](http://stackoverflow.com/questions/1083676/understanding-how-join-works-when-3-or-more-tables-are-involved-sql)
The first answer explains how it works conceptually:

- The optimizer comes up with a plan, which includes a join order. It could be A, B, C, or C, B, A or any of the combinations
- The query execution engine applies any predicates (WHERE clause) to the first table that doesn't involve any of the other tables. It selects out the columns mentioned in the JOIN conditions or the SELECT list or the ORDER BY list. Call this result A
- It joins this result set to the second table. For each row it joins to the second table, applying any predicates that may apply to the second table. This results in another temporary resultset.
- Then it joins in the final table and applies the ORDER BY

The article mentions the usage of FILTERS, which we haven't discusses so far, I think... but it doesn't affect the validity of the JOIN explanation.

Does it help?
